OpenRA.Game/Traits/World/Shroud.cs

```diff
@@ -163,42 +166,105 @@ void UpdateVisibility(Actor a, ref CPos[] visible)
 			AddVisibility(a, ref visible);
 		}
 
-		void AddShroudGeneration(Actor a, ref CPos[] shrouded)
+		void AddDisruptionGenerator(Actor a, ref CPos[] underDisruption)
+		{
+			var cd = a.TraitOrDefault<CreatesDisruptionField>();
+			if (cd == null || a.Owner.IsAlliedWith(self.Owner) || cd.Range == WRange.Zero || !lobbyFogEnabled)
+				return;
+
+			// Lazily generate the disrupted tiles, allowing the caller to re-use them if desired.
+			underDisruption = underDisruption ?? FindVisibleTiles(a, cd.Range);
+
+			foreach (var c in underDisruption)
+				generatedDisruptionCount[c]++;
+
+			if (disruptionGenerators.ContainsKey(a))
+				throw new InvalidOperationException("Attempting to add duplicate shroud generation");
```

Comment:
If you are going to change logic please update any accompanying messages.
This is now *disruption*, not shroud.